File: LeetCode-POTD-Solutions/Delete-Node-in-a-BST.java
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) return null;

        if (key < root.val) {
            root.left = deleteNode(root.left, key);
        } 
        else if (key > root.val) {
            root.right = deleteNode(root.right, key);
        } 
        else {
            // Node found

            // Case 1: No left child
            if (root.left == null) return root.right;

            // Case 2: No right child
            if (root.right == null) return root.left;

            // Case 3: Both children -> find inorder successor
            TreeNode cur = root.right;
            while (cur.left != null) {
                cur = cur.left;
            }

            root.val = cur.val;  // Replace value
            root.right = deleteNode(root.right, cur.val); // Delete successor
        }

        return root;
    }
}

File: LeetCode-POTD-Solutions/Delete-Node-in-a-BST.py
class Solution(object):
    def deleteNode(self, root, key):

        if not root:
            return None

        if key < root.val:
            root.left = self.deleteNode(root.left, key)

        elif key > root.val:
            root.right = self.deleteNode(root.right, key)

        else:
            # Case 1: No left child
            if not root.left:
                return root.right

            # Case 2: No right child
            if not root.right:
                return root.left

            # Case 3: Two children → find inorder successor
            cur = root.right
            while cur.left:
                cur = cur.left

            root.val = cur.val
            root.right = self.deleteNode(root.right, cur.val)

        return root
